File: algorithm/c++/TheGame1516/theGame1516.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>
#include <queue>

using namespace std;

int N;
vector<int> ver[501];
vector<int> adj[501];
queue<int> que;

int times[501];
int result[501];
int in_deg[501];

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);
    
    cin >> N;
    
    for(int i = 1; i <= N; i++)
    {
        int input;
        cin >> input;

        while(input != -1)
        {
            ver[i].push_back(input);
            cin >> input;
        }
    }

    for (int i = 1; i <= N; i++)
    {
        times[i] = ver[i][0];
    }

    for (int i = 1; i <= N; i++)
    {
        for (int j = 1; j < ver[i].size(); j++)
        {
            adj[ver[i][j]].push_back(i);
            in_deg[i]++;
        }
    }

    for (int i = 1; i <= N; i++)
    {
        if (in_deg[i] == 0) que.push(i);
        result[i] = times[i];
    }

    
    while(!que.empty())
    {
        int cur = que.front();
        que.pop();

        for (int i = 0; i < adj[cur].size(); i++)
        {
            int next = adj[cur][i];
            in_deg[next]--;

            result[next] = max(result[next], result[cur] + times[next]);

            if (in_deg[next] == 0) que.push(next);
        }
    }

    for (int i = 1; i <= N; i++) cout << result[i] << endl;


    return 0;
}

File: algorithm/python/TheGame1516/theGame1516.py
import sys
from queue import Queue

input = sys.stdin.readline

N = int(input())

adj = [[] for _ in range(N+1)]
vertex = [[] for _ in range(N+1)]

times = [0] * (N+1)
result = [0] * (N+1)
degree = [0] * (N+1)

for i in range(1,N+1):
    data = list(map(int, input().split()))[:-1]
    for j in range(0,len(data)):
        vertex[i].append(data[j])
    
for i in range(1, N+1):
    times[i] = vertex[i][0]
    result[i] = vertex[i][0]

for i in range(1, N+1):
    for j in range(1, len(vertex[i])):
        adj[vertex[i][j]].append(i)
        degree[i] = degree[i] + 1

queue = Queue()

for i in range(1, N+1):
    if (degree[i] == 0):
        queue.put(i)

while not queue.empty():

    cur = queue.get()

    for i in range(0, len(adj[cur])):
        next = adj[cur][i]
        degree[next] = degree[next] - 1

        result[next] = max(result[next], result[cur] + times[next])

        if (degree[next] == 0):
            queue.put(next)

for i in range(1,N+1):
    print(result[i])